src/loader/mod.rs

```diff
@@ -0,0 +1,580 @@
+/// * `kernel_end` - The offset of `guest_mem` where kernel image load is loaded finish, return
+///                  in case of loading initrd adjacent to kernel image.
+/// * `setup_header` - The setup_header belongs to linux boot protocol, only for bzImage, vmm
+///                    will use it to setup setup_header.init_size, which is a must for bzImage
+///                    direct boot.
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
+pub struct KernelLoaderResult {
+    pub kernel_load: GuestAddress,
+    pub kernel_end: GuestUsize,
+    pub setup_header: Option<bootparam::setup_header>,
+}
+
+pub trait KernelLoader {
+    fn load<F>(
+        guest_mem: &GuestMemoryMmap,
+        kernel_start: Option<GuestAddress>,
+        kernel_image: &mut F,
+        lowest_kernel_start: Option<GuestAddress>,
+    ) -> Result<(KernelLoaderResult)>
+    where
+        F: Read + Seek;
+}
+
+pub struct Elf;
+
+impl KernelLoader for Elf {
+    /// Loads a kernel from a vmlinux elf image to a slice
+    ///
+    /// kernel is loaded into guest memory at offset phdr.p_paddr specified by elf image.
+    ///
+    /// # Arguments
+    ///
+    /// * `guest_mem` - The guest memory region the kernel is written to.
+    /// * `kernel_start` - The offset into 'guest _mem' at which to load the kernel.
+    /// * `kernel_image` - Input vmlinux image.
+    /// * `lowest_kernel_start` - This is the start of the high memory, kernel should above it.
+    ///
+    /// # Returns
+    /// * KernelLoaderResult
+    fn load<F>(
+        guest_mem: &GuestMemoryMmap,
+        kernel_start: Option<GuestAddress>,
+        kernel_image: &mut F,
+        lowest_kernel_start: Option<GuestAddress>,
+    ) -> Result<(KernelLoaderResult)>
+    where
+        F: Read + Seek,
+    {
+        let mut ehdr: elf::Elf64_Ehdr = Default::default();
+        kernel_image
+            .seek(SeekFrom::Start(0))
+            .map_err(|_| Error::SeekElfStart)?;
+        unsafe {
+            // read_struct is safe when reading a POD struct.  It can be used and dropped without issue.
+            struct_util::read_struct(kernel_image, &mut ehdr).map_err(|_| Error::ReadElfHeader)?;
+        }
+
+        // Sanity checks
+        if ehdr.e_ident[elf::EI_MAG0 as usize] != elf::ELFMAG0 as u8
+            || ehdr.e_ident[elf::EI_MAG1 as usize] != elf::ELFMAG1
+            || ehdr.e_ident[elf::EI_MAG2 as usize] != elf::ELFMAG2
+            || ehdr.e_ident[elf::EI_MAG3 as usize] != elf::ELFMAG3
+        {
+            return Err(Error::InvalidElfMagicNumber);
+        }
+        if ehdr.e_ident[elf::EI_DATA as usize] != elf::ELFDATA2LSB as u8 {
+            return Err(Error::BigEndianElfOnLittle);
+        }
+        if ehdr.e_phentsize as usize != mem::size_of::<elf::Elf64_Phdr>() {
+            return Err(Error::InvalidProgramHeaderSize);
+        }
+        if (ehdr.e_phoff as usize) < mem::size_of::<elf::Elf64_Ehdr>() {
+            // If the program header is backwards, bail.
```

Comment:
What do you mean by backwards here?

Reply:
means overlap. program header should follow after elf header,  it's invalid if not, that's what the above adjustment presents.

Reply:
Can you rephrase it so it's more clear what you want to check here? Program headers is backwards doesn't say much to me. Any reason you added a comment to this check, but didn't added comments to the other ones?